### PR TITLE
Better processing of strings with newlines

### PR DESCRIPTION
**Pull request by aidan-plenert-macdonald**
Fixes https://github.com/google/protobuf/issues/2798

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by aidan-plenert-macdonald** [comment]
I signed it!

My company did it for Tensorflow. Do I need to do it again?

**Comment by acozzette** [comment]
ok to test (these magic words will kick off the CI builds)

@aidan-plenert-macdonald Thanks for the pull request! I believe a CLA for TensorFlow should work here, too; are you sure the email address on your commit matches the one associated with your company and the CLA?

**Comment by aidan-plenert-macdonald** [comment]
I am pretty sure it matches. I think I originally committed with my knupath email but amended the commit with my knuedge email (I have lots of emails). I am pretty sure the knuedge email is the right one.

The test failed. Looks like my Regex is consuming the quote. `ParseError: 1:226 : String missing ending quote: '"\\303\\274\\352\\234\\23`

If you tell me how to run the tests locally, I will fix them

**Comment by acozzette** [comment]
To run the tests, you should be able to run `python setup.py build && python setup.py test` within the `python/` subdirectory.

About the CLA, I think your knuedge.com email is OK and present on the commit but you may need to add it to your GitHub profile (it doesn't have to be the primary email address). Could you try that and see if it works?

**Comment by aidan-plenert-macdonald** [comment]
Thanks. That email is there. What else should I try?

**Comment by acozzette** [comment]
Ok, thanks, actually now I think the problem is that somehow the CLA system does not seem to know that your email address is associated with the KnuEdge CLA; let me investigate a bit and get back to you.

**Comment by aidan-plenert-macdonald** [comment]
@acozzette  It appears to pass all the tests when run locally. Do I need to run the skipped ones? How?

{code}
protobuf/python$ python setup.py build && python setup.py test 2>&1 | tail -3
running build
running build_py
Ran 604 tests in 0.505s

OK (skipped=7)
{code}

**Comment by acozzette** [comment]
retest this please (this will trigger the CI bots again)

@aidan-plenert-macdonald No need to worry about the skipped tests.

**Comment by aidan-plenert-macdonald** [comment]
I signed it!

**Comment by aidan-plenert-macdonald** [comment]
@acozzette Can you check the CLA again? I got added to my companies group for the CLA. https://groups.google.com/a/knuedge.com/forum/#!forum/tensorflow_developers

**Comment by aidan-plenert-macdonald** [comment]
I signed it!

**Comment by aidan-plenert-macdonald** [comment]
I signed it!

**Comment by acozzette** [comment]
retest this please (that phrase should trigger the CI bots to run again)

File: python/google/protobuf/text_format.py [hunk 1]
```diff
@@ -469,7 +469,7 @@ def Merge(text,
     ParseError: On text parsing problems.
   """
   return MergeLines(
-      text.split('\n'),
+      re.findall('(?:[^\"\n]|(?:\"(?:[^\"]|[\n])*\"))*\n?', text),
```
Comment (acozzette) [inline python/google/protobuf/text_format.py]: I think we would want to make the same change in the Parse() method above for consistency.

Could you perhaps also add a comment that breaks down what that regex is doing?

Comment (aidan-plenert-macdonald) [inline python/google/protobuf/text_format.py]: Yeah. Making people read Regex is a pretty harsh punishment. I added it. Hope it makes sense.

**Comment by acozzette** [comment]
Jenkins is showing some new test failures that look related, though I haven't looked closely yet at why they're failing: https://grpc-testing.appspot.com/job/protobuf_pull_request/1115/

**Comment by aidan-plenert-macdonald** [comment]
Yeah. That happened before. I mentioned above that I can't replicate the test cases locally to examine them. 

https://github.com/google/protobuf/pull/2807#issuecomment-286160376

**Comment by acozzette** [comment]
So you can run the tests locally and they all just pass already?

**Comment by aidan-plenert-macdonald** [comment]
I might be missing tests hat Jenkins has, but yeah. Mine passed fine

**Comment by acozzette** [comment]
It looks like for some reason the test is only failing for the Python C++ implementation. When I try it I see the failure if I run this: `python setup.py build --cpp_implementation && python setup.py test --cpp_implementation`

**Comment by aidan-plenert-macdonald** [comment]
I can't figure out how to get rid of this error.

```
$ python setup.py build --cpp_implementation && python setup.py test --cpp_implementation
running build
running build_py
running build_ext
running test
running egg_info
writing requirements to protobuf.egg-info/requires.txt
writing protobuf.egg-info/PKG-INFO
writing namespace_packages to protobuf.egg-info/namespace_packages.txt
writing top-level names to protobuf.egg-info/top_level.txt
writing dependency_links to protobuf.egg-info/dependency_links.txt
reading manifest file 'protobuf.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files found matching 'google/protobuf/internal/*.proto'
warning: no previously-included files matching '*_test.py' found under directory 'google'
warning: no previously-included files matching '*_test.proto' found under directory 'google'
warning: no previously-included files matching '*.dll' found anywhere in distribution
warning: no previously-included files matching '*.pyc' found anywhere in distribution
warning: no previously-included files matching '*.pyo' found anywhere in distribution
warning: no previously-included files matching '*.so' found anywhere in distribution
writing manifest file 'protobuf.egg-info/SOURCES.txt'
running build_ext
copying build/lib.linux-x86_64-2.7/google/protobuf/pyext/_message.so -> google/protobuf/pyext
copying build/lib.linux-x86_64-2.7/google/protobuf/internal/_api_implementation.so -> google/protobuf/internal
Traceback (most recent call last):
  File "setup.py", line 260, in <module>
    ext_modules=ext_module_list,
  File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/usr/local/lib/python2.7/dist-packages/setuptools/command/test.py", line 211, in run
    self.run_tests()
  File "/usr/local/lib/python2.7/dist-packages/setuptools/command/test.py", line 234, in run_tests
    **exit_kwarg
  File "/usr/lib/python2.7/unittest/main.py", line 94, in __init__
    self.parseArgs(argv)
  File "/usr/lib/python2.7/unittest/main.py", line 149, in parseArgs
    self.createTests()
  File "/usr/lib/python2.7/unittest/main.py", line 158, in createTests
    self.module)
  File "/usr/lib/python2.7/unittest/loader.py", line 130, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/lib/python2.7/unittest/loader.py", line 103, in loadTestsFromName
    return self.loadTestsFromModule(obj)
  File "/usr/local/lib/python2.7/dist-packages/setuptools/command/test.py", line 43, in loadTestsFromModule
    tests.append(self.loadTestsFromName(submodule))
  File "/usr/lib/python2.7/unittest/loader.py", line 100, in loadTestsFromName
    parent, obj = obj, getattr(obj, part)
AttributeError: 'module' object has no attribute 'service_reflection_test'
```

**Comment by acozzette** [comment]
Does it help if you run it in virtualenv, like this?

```
virtualenv env
source env/bin/activate
export LD_LIBRARY_PATH=$PWD/../src/.libs
python setup.py build --cpp_implementation && python setup.py test --cpp_implementation
```

I think last time I saw an error like that it was because the tests were mixing up the local protobuf with the one installed through my system's package manager.

**Comment by aidan-plenert-macdonald** [comment]
Yeah. Thanks. I'll fix the bug

**Comment by anandolee** [comment]
Closing it for no fix more than one year. Feel free to reopen if you are still working on it

**Pull request closed without merging**